### PR TITLE
Fix #997

### LINKED ISSUE #997: Private types should not be exposed by non-private members
We should not allow a private type to appear in the signature of a non-private member. 
The private type is not accessible outside the scope of the member. In a "by-name" based
system for resolving types this can lead to crashes. 

Example: `collection.mutable.IndexedSeqView` crashes when the accessibility of the `This` type
is restricted from `protected[this]` to `private[this]`. The crash happens in mixin when we try to refer
to `This` computing the signature of `filter` in `AbstractTransformedX`, which is an inner subclass
of `IndexedSeqView`. It can't work because `This` is not visible as an inherited member of `AbstractTransformedX`. To prevent crashes like these we have to be stricter in the frontend.

### PR DESCRIPTION
**Pull request by odersky**
Ensure that private references do not leak. Review by @DarkDimius

### CODE REVIEW

File: src/dotty/tools/dotc/typer/Checking.scala [hunk 1]
```diff
@@ -296,6 +296,46 @@ object Checking {
...
+        if (sym.is(Private)) sym.owner
+        else if (sym.privateWithin.exists) sym.privateWithin
+        else if (sym.is(Package)) sym
+        else accessBoundary(sym.owner)
+      def apply(errors: Errors, tp: Type): Errors = tp match {
+        case tp: NamedType =>
+          val errors1 =
+            if (tp.symbol.is(Private) &&
+               !accessBoundary(sym).isContainedIn(tp.symbol.owner)) {
+              (d"non-private $sym refers to private ${tp.symbol}\n in its type signature ${sym.info}", tree.pos) :: errors
```
Comment (smarter) [inline src/dotty/tools/dotc/typer/Checking.scala]: Why not just call `ctx.errorOrMigrationWarning` here?

Comment (odersky) [inline src/dotty/tools/dotc/typer/Checking.scala]: Because we might end up not issuing the error if a parent is a TypeAlias.

**Comment by odersky** [comment]
Ready to merge?

**Comment by DarkDimius** [comment]
LGTM

**Comment by odersky** [comment]
Rebased to master

**Pull request merged**